contracts/subdomainregistrar/BaseSubdomainRegistrar.sol

```diff
@@ -0,0 +1,263 @@
+            revert Unauthorised(node);
+        }
+        _;
+    }
+    modifier canBeRegistered(bytes32 parentNode, uint64 duration) {
+        _checkParent(parentNode, duration);
+        _;
+    }
+
+    function available(bytes32 node) public view virtual returns (bool) {
+        try wrapper.getData(uint256(node)) returns (
+            address,
+            uint32,
+            uint64 expiry
+        ) {
+            return expiry < block.timestamp;
+        } catch {
+            return true;
+        }
+    }
+
+    function _setupDomain(
+        bytes32 node,
+        ISubdomainPricer pricer,
+        address beneficiary,
+        bool active
+    ) internal virtual authorised(node) {
+        names[node] = Name({
+            pricer: pricer,
+            beneficiary: beneficiary,
+            active: active
+        });
+        emit NameSetup(node, address(pricer), beneficiary, active);
+    }
+
+    function _batchRegister(
+        bytes32 parentNode,
+        string[] calldata labels,
+        address[] calldata addresses,
+        address resolver,
+        uint16 fuses,
+        uint64 duration,
+        bytes[][] calldata records
+    ) internal {
+        if (
+            labels.length != addresses.length || labels.length != records.length
+        ) {
+            revert DataMissing();
+        }
+
+        if (!names[parentNode].active) {
+            revert ParentNameNotSetup(parentNode);
+        }
+
+        _checkParent(parentNode, duration);
+
+        _batchPayBeneficiary(parentNode, labels, duration);
+
+        //double loop to prevent re-entrancy because _register calls user supplied functions
+
+        for (uint256 i = 0; i < labels.length; i++) {
+            _register(
+                parentNode,
+                labels[i],
+                addresses[i],
+                resolver,
+                fuses,
+                uint64(block.timestamp) + duration,
+                records[i]
+            );
+        }
+    }
+
+    function register(
+        bytes32 parentNode,
+        string calldata label,
+        address newOwner,
+        address resolver,
+        uint32 fuses,
+        uint64 duration,
+        bytes[] calldata records
+    ) internal {
+        if (!names[parentNode].active) {
+            revert ParentNameNotSetup(parentNode);
+        }
+
+        (address token, uint256 fee) = ISubdomainPricer(
+            names[parentNode].pricer
+        ).price(parentNode, label, duration);
+
+        _checkParent(parentNode, duration);
+
+        if (fee > 0) {
+            IERC20(token).transferFrom(
+                msg.sender,
+                address(names[parentNode].beneficiary),
+                fee
+            );
+        }
+
+        _register(
+            parentNode,
+            label,
+            newOwner,
+            resolver,
+            fuses,
+            uint64(block.timestamp) + duration,
+            records
+        );
+    }
+
+    /* Internal Functions */
```

Comment:
The `register` function at l120 is also set as `internal` so this comment should come above l120 (unless you mistakenly marked `register` as `internal` when it was meant for `public`

Reply:
Maybe better to rename as `_singleRegister` to be consistent with `_batchRegister`